openstack_sdk/src/config.rs

```diff
@@ -358,51 +461,103 @@ fn get_config_file_search_paths<S: AsRef<str> + std::fmt::Display>(filename: S)
+/// - `/etc/openstack/clouds.{yaml,yml,json}`
+pub fn find_clouds_file() -> Option<PathBuf> {
+    get_config_file_search_paths("clouds")
+        .into_iter()
+        .find(|path| path.is_file())
+}
+
+/// Searches for a `secure.{yaml,yml,json}` config file.
+///
+/// The following locations will be tried in order:
+///
+/// - `./secure.{yaml,yml,json}` (current working directory)
+/// - `$XDG_CONFIG_HOME/openstack/secure.{yaml,yml,json}`
+/// - `$XDG_HOME/.config/openstack/secure.{yaml,yml,json}`
+/// - `/etc/openstack/secure.{yaml,yml,json}`
+pub fn find_secure_file() -> Option<PathBuf> {
+    get_config_file_search_paths("secure")
+        .into_iter()
+        .find(|path| path.is_file())
 }
 
 impl ConfigFile {
-    /// Get new ConfigFile processor with default config file search locations.
-    pub fn new() -> Result<Self, ConfigError> {
-        let mut s = config::Config::builder();
-        for filename in &["clouds", "secure", "clouds-public"] {
-            if let Some(path) = get_config_file_search_paths(filename)
-                .iter()
-                .find(|&x| x.is_file())
-            {
-                if let Some(v) = path.to_str() {
-                    match is_config_valid(&path) {
-                        Ok(true) => s = s.add_source(File::with_name(v)),
-                        Ok(false) => error!("Ignoring file {}", v),
-                        Err(x) => return Err(x),
-                    }
+    /// A builder to create a `ConfigFile` by specifying which files to load.
+    pub fn builder() -> ConfigFileBuilder {
+        ConfigFileBuilder {
+            sources: Vec::new(),
+        }
+    }
+
+    /// Create a `ConfigFile` which also loads the default sources in the following order:
+    ///
+    /// - `clouds-public.{yaml,yml,json}` (see [`find_vendor_file`] for search paths)
+    /// - `clouds.{yaml,yml,json}` (see [`find_clouds_file`] for search paths)
+    /// - the provided clouds file
+    /// - `secure.{yaml,yml,json}` (see [`find_secure_file`] for search paths)
+    /// - the provided secure file
+    ///
+    /// If a source has syntax errors (YAML/JSON) or the file does not exist it will be ignored,
+    /// but if one of the default sources is not a valid config, a [`ConfigError`] will be returned.
+    pub fn new_with_user_specified_configs(
+        clouds: Option<impl AsRef<Path>>,
+        secure: Option<impl AsRef<Path>>,
+    ) -> Result<Self, ConfigError> {
+        let mut builder = Self::builder();
+
+        for path in find_vendor_file()
+            .into_iter()
+            .chain(find_clouds_file())
+            .chain(clouds.map(|path| path.as_ref().to_owned()))
+            .chain(find_secure_file())
+            .chain(secure.map(|path| path.as_ref().to_owned()))
+        {
+            builder = match builder.add_source(path) {
+                Ok(builder) => builder,
+                Err(ConfigFileBuilderError::FileParse { path, builder, .. }) => {
```

Comment:
it took me a while to understand why the behavior did change. You actually swapped handling of config errors:

Previous (current code)
- when it is not possible to load the file (i.e. broken yaml) - error is thrown (ConfigError::parse)
- when file cannot be deserialized (i.e. string where object is expected) - error is logged (pointing to the exact error in the file) and the file is skipped. In this case most likely the auth will fail

new:
- if the file can not be loaded (broken yaml) it is skipped => auth fails
- when the file cannot be deserialized - an error is raised.

So, to keep the current behavior please swap branches and include deserialization error into the error message (`openstack_sdk::config: Error in file /home/USER/.config/openstack/secure.yaml: invalid type: string "foo", expected a map. Ignoring the file`)

Afterwards we can maybe play around with options whether we want to ignore errors or stop on first error, but I would like to get the change first since it is allowing better tests

Reply:
Sorry wasn't intentional. I have changed it. Hopefully it now is correct.

Reply:
Sure, no worries :)